cpp/dolfinx/io/XDMFFile.cpp

```diff
@@ -363,10 +382,83 @@ XDMFFile::read_meshtags(const mesh::Mesh<double>& mesh, std::string name,
                                + "' not found.");
     }
     else
+    {
       values_data_node = attribute_node.child("DataItem");
+    }
   }
-  const std::vector values = xdmf_utils::get_dataset<std::int32_t>(
-      _comm.comm(), values_data_node, _h5_id);
+
+  // Now we need to figure out what type of data the file contains and check
+  // that it is compatible with what we are reading.
+
+  // If the data is stored in HDF format
+  if (const auto& format_attribute = values_data_node.attribute("Format");
+      format_attribute && std::string(format_attribute.value()) == "HDF")
+  {
+    const auto& [file_name, node_path]
+        = xdmf_utils::get_hdf5_paths(values_data_node);
+
+    // Then we need to read the data type from the .h5 file
+    const auto datatype_id
+        = io::hdf5::get_dataset_datatype(_h5_id, node_path.c_str());
+
+    if (!H5Tequal(datatype_id, io::hdf5::hdf5_type<T>()))
+    {
+      throw std::runtime_error(
+          "The data in the XDMF file does not match the required data type.");
+    }
+  }
+  else
+  {
+    // Otherwise, this information should be stored in node attributes in the
+    // XML file
+    std::string data_type;
+
+    // "DataType" is the attribute used in the current version of the
+    // standard...
+    if (const auto& data_type_attribute
+        = values_data_node.attribute("DataType");
+        data_type_attribute)
+    {
+      data_type = data_type_attribute.value();
+    }
+    // ... But "NumberType" is supported too and widely used in its place
+    else if (const auto& data_type_attribute
+             = values_data_node.attribute("NumberType");
+             data_type_attribute)
+    {
+      data_type = data_type_attribute.value();
+    }
+    // If unspecified, use the standard's implicit default
+    else
+    {
+      data_type = "Float";
+    }
+
+    int precision;
+
+    // If there is an attribute that specifies the precision, read it from there
+    if (const auto& precision_attribute
+        = values_data_node.attribute("Precision");
+        precision_attribute)
+    {
+      precision = std::stol(values_data_node.attribute("Precision").value());
+    }
+    // otherwhise use the standard's default
+    else
+    {
+      precision = 4;
+    }
```

Comment:
```suggestion
      precision = std::stol(values_data_node.attribute("Precision").value());
    // otherwhise use the standard's default
    else
      precision = 4;
```
I think this would be the "house-style"